docs/debugging/debugging.md

```diff
@@ -0,0 +1,205 @@
+    like this, reformatted here for clarity:
+
+```
+INFO [2021-12-01 09:26:26] 
+  <sending-ip-address> "Faraday v1.8.0"
+  <host-ip-address>:8000
+  GET /editorcheck
+    ?bot_name=ropensci-review-bot&
+     issue_author=<author>&
+     issue_id=<n>1&
+     repo=ropensci%2Fsoftware-review&
+     repourl=https%3A%2F%2Fgithub.com%2F<org>%2F<repo>&
+     sender=ropensci-review-bot
+  200
+  1.235
+```
+
+These logs can be used to examine commands issued by the bot and passed on by
+the Heroku serive, which always have the machine information "Faraday".
+
+:::{admonition} Debug Step 3
+:class: tip
+Ensure that the command was received and logged by the `roreviewapi` service.
+:::
+
+
+## The `/stdlogs` endpoint
+
+Most commands, including `check package` and `check srr`, issued either
+manually or automatically on all new submissions return immediately with a
+message, while starting a background process for the actual package checks. The
+`\stdlogs` endpoint is intended to help diagnose issues with this background
+checking process.
+
+This background process dumps all `stdout` and `stderr` messages to cached log
+files. The contents of these can be accessed either as HTTP requests to the
+running process. This is a `GET` endpoint with the single parameter of
+`repourl` specifying the URL of the package being checked. Logs are cached only
+for the most recent calls for any one package, and are over-written by any
+subsequent calls.
+
+It is nevertheless often easier to directly log in to the running processes via
+SSH as described above, in which case the logs are dumped within a `templogs`
+sub-directory of `/home/.cache/R/pkgcheck/`. These log files are all prefixed
+with the name of the package being checked.
+
+:::{admonition} Debug Step 4
+:class: tip
+Ensure that log files were generated for the package.
+
+- If using the HTTP endpoints, calls should always succeed as long as log files
+exist.
+- If using SSH, there should be two files prefixed with the package name.
+:::
+
+If the bot has entirely failed to respond, it is likely because of an internal
+error, in which case the `stderr` log should hopefully contain sufficient
+information to help debug the problem. The `stdout` log may provide additional
+detail in some cases.
+
+If these logs confirm an internal error, the next step is generally to try to
+reproduce that error locally, through cloning the package and calling, or
+stepping through, the relevant [endpoint
+function](/roreviewapi/vignettes/endpoints).
+
+:::{admonition} Debug Step 5
+:class: tip
+Try to locally reproduce any errors in the `stderr` log files, and solve
+problem from there.
+:::
+
+## Manual Debugging
+
+Some errors may arise in which a `log` entry appears to have been correctly
+sent, and yet no `stdlogs` are generated. These typically require manual
+debugging. The best approach is then to manually run the `roreviewapi` Docker
+container on the Digital Ocean server, and step through the code to locate the
```

Comment:
I'll need a demo one day of this.

Reply:
I have no idea how to demo that :confused: It used to happen very infrequently, but hasn't happened for years, at least, so (1) I've no idea how to trigger it; and (2) I presume it'll be very unlikely to happen again in the future.